### DISCUSSION TITLE
add a general understanding section in README.md

### DISCUSSION THREAD

**Discussion by FuPeiJiang** [WoeUSB/WoeUSB-ng#30] [Ideas]
(use UEFI if you can)

### UEFI: no MBR involved
UEFI looks for `.EFI` files in fat(32/16/12) partition.
your Windows `.EFI` is in NTFS partition (UEFI cannot see it)
we're going to make a fat32 partition with `.EFI` that will chainload Windows `.EFI` (the chainloader `.EFI` can see Windows `.EFI`)

1. we put `.EFI` files from [UEFI:NTFS](https://github.com/pbatard/uefi-ntfs) in the fat32 partition
we download from [here](https://github.com/pbatard/rufus/raw/master/res/uefi/uefi-ntfs.img), it's an `.img`, if you mount it, you'll see `.EFI` files
2. when you boot using `.EFI` from UEFI:NTFS<br> UEFI:NTFS will find the first Windows partition or NTFS partition<br>and chainload `.EFI`: for x64 systems, it's `/efi/boot/bootx64.efi`

### BIOS: no .EFI file involved
BIOS looks in the MBR to see what to boot

1. GRUB writes to the MBR to make it boot into GRUB
2. normally GRUB has a menu (Ex: [Ubuntu Menu](https://www.omgubuntu.co.uk/wp-content/uploads/2012/09/grub2-in-ubuntu.jpg)), but we tell GRUB to directly boot into windows
by making `grub.cfg` exactly as such:
```
ntldr /bootmgr
boot
```
___
BIOS is bad: MBR can only tell BIOS to boot one operating system, MBR telling BIOS to boot into GRUB let's us use many operating systems. 
UEFI looks for .EFI files, if you have many .EFI files, you can have many boot options listed in UEFI boot order menu.
though having .EFI in NTFS doesn't help, because UEFI doesn't see NTFS. 

some motherboards provide drivers to see NTFS: in this case, you don't need the fat32 partition, it sees the .EFI in the NTFS Windows Partition.

**Comment by FuPeiJiang** [comment]
I think every project needs a general understanding section in README.md, to help new contributors understand, well, here it wasn't really needed because the python code is SOOOOOOOOOOOOOO clean. tysm, I was really confused about the bootable magic, was doing tests, found this repo and it helped me understand, I don't think I could've otherwise
___
also, `./WoeUSB/core.py` is the main stuff
___
I had to move
 `./WoeUSB/woeusb` to `./woeusb`
 `./WoeUSB/woeusbgui` to `./woeusbgui`

or else it said module not found (I don't remember the exact error message, I'm on Windows now)
I was running directly from source.

should we move these ?

**Reply by WaxyMocha** [reply]
Hi, I know that package files need to be in WoeUSB/, but since `woeusb` and `woeusbgui` are now just bootstraps, I think we could move it outside if it works better that way.

**Reply by WaxyMocha** [reply]
About general understanding, it might be a good idea, but I would do it in `Wiki` and just link it in README.

**Reply by brlin-tw** [reply]
Or in <https://github.com/WoeUSB/WoeUSB/wiki> since the content shall be the same for both projects.

I've created [WoeUSB boot process explained · WoeUSB/WoeUSB Wiki](https://github.com/WoeUSB/WoeUSB/wiki/WoeUSB-boot-process-explained) just for a start.

**Reply by brlin-tw** [reply]
~~I doubt whether we should add the link to the README since people should be able to prepare an installer drive without knowing the underlying details...~~

Maybe it should be added in the "How to use <product>" section of the README as an auxiliary material.

**Reply by FuPeiJiang** [reply]
it is totally not "How to use"
it's to help people become contributors
`CONTRIBUTING.md` then, but I fear no one reads that.

so better link to `GENERAL UNDERSTANDING` in README.md
it'd be good to help users understand the process (if they want to)
maybe gain more contributors

**Reply by FuPeiJiang** [reply]
I think we should put my explanation as LEVEL1, and show it first.
after someone understands that, they can understand LEVEL2. (I'd rank this as LEVEL2.5)

[LEVEL2.5](https://github.com/WoeUSB/WoeUSB/wiki/WoeUSB-boot-process-explained) is too long and doesn't hightlight the essential, so it's hard to grasp the essential

normally I'm only interested in what I need

so:
general understanding LEVEL1
WoeUSB boot process explained LEVEL2

**Comment by FuPeiJiang** [comment]
yeah... I think we shouldn't put it in the wiki because I totally missed it.
nothing points to the wiki in `README.md` or I missed that too

a link to `general understanding` should be in README.md or CONTRIBUTING.md
because normally the wiki is empty (so no one looks there), and it isn't a good place to put documentation/wiki. (it's not searchable)

my favorite documentation is https://www.autohotkey.com/docs/AutoHotkey.htm
[here's the source](https://github.com/Lexikos/AutoHotkey_L-Docs)
it's searchable
you can even search it this way:
https://www.autohotkey.com/docs/commands/{your search query}.htm
https://www.autohotkey.com/docs/commands/==.htm
https://www.autohotkey.com/docs/commands/array.htm
___
where are the files to the wiki located ? can it only be edited on github ?
I can't PR to the wiki. I suggest creating a repo for the wiki, or putting the wiki in a folder in the main repo.

**Comment by FuPeiJiang** [comment]
GENERAL UNDERSTANDING

could help people do/write variations
for example: I want to put Ubuntu And Windows in one USB. and I've finally managed to do that.

you don't need [UEFI:NTFS](https://github.com/pbatard/uefi-ntfs)

you can use .EFI from GRUB and GRUB can do the chainloading:

```
menuentry "Windows 10" {
    insmod ntfs
    insmod part_msdos
    insmod search_fs_uuid
    insmod chain
    insmod part_gpt
    search --set=root --label windows --hint hd0,msdos2
    chainloader /efi/boot/bootx64.efi
}
```
of course this GRUB is for Ubuntu. But with this menuentry it can also boot Windows

___
___
explaining the function of code
not the implementation.
could help people find better implementations

so a file explaining function: not even pseudocode.
if I write that, could it be added to the project?
